### PR TITLE
SD-330: slice admin campaign page

### PR DESCRIPTION
**Pull request by domagojlatecki**

### CODE REVIEW

File: src/app/admin/admin-campaign-new/admin-campaign-new.component.ts [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+      }
+    }
+
+    return null
+  }
+
+  private static validDateRange(formGroup: FormGroup): ValidationErrors | null {
+    if (formGroup.value.startDate >= formGroup.value.endDate) {
+      return {invalidDateRange: true}
+    }
```
Comment (mightymatth) [inline src/app/admin/admin-campaign-new/admin-campaign-new.component.ts:291]: Isn't `startDate == endDate` a valid input? Theoretically, it is possible to have a one-day period for a project. 

image
😆

File: src/app/admin/admin-ft-asset-detail/admin-ft-asset-detail.component.html [hunk 2]
```diff
@@ -5,7 +5,7 @@
         <app-admin-asset-item [asset]="assetRes.value" assetType="asset" type="edit-screen"></app-admin-asset-item>
 
         <div class="flex flex-col lg:flex-row justify-center items-center gap-4 my-8">
-          <app-add-button text="NEW CAMPAIGN"></app-add-button>
+          <app-add-button text="NEW CAMPAIGN" [routerLink]="'campaigns/new'"></app-add-button>
```
Comment (mightymatth) [inline src/app/admin/admin-ft-asset-detail/admin-ft-asset-detail.component.html:8]: `routerLink="campaigns/new"`?

**Review by mightymatth — CHANGES_REQUESTED** [review]

File: src/app/admin/admin-campaign-new/admin-campaign-new.component.ts [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+      hardCapTokensPercentage: [0, Validators.required],
+      tokenPrice: [0, Validators.required],
+      hasMinAndMaxInvestment: [false, Validators.required],
+      minInvestment: [{value: undefined, disabled: true}, Validators.required],
+      maxInvestment: [{value: undefined, disabled: true}, Validators.required],
+      isReturningProfitsToInvestors: [false, Validators.required],
+      returnFrequency: [{value: '', disabled: true}, Validators.required],
+      isReturnValueFixed: [false, Validators.required],
+      returnFrom: [{value: undefined, disabled: true}, [Validators.required, Validators.min(0.01), Validators.max(1)]],
+      returnTo: [{value: undefined, disabled: true}, [Validators.required, Validators.min(0.01), Validators.max(1)]],
```
[the remark this answers is not in the document]
Comment (domagojlatecki) [reply src/app/admin/admin-campaign-new/admin-campaign-new.component.ts:60]: Do you mean nested form groups or just separate form groups altogether? Currently I grouped everything into two form groups depending on the creation step which is displayed.

File: src/app/admin/admin-campaign-new/admin-campaign-new.component.ts (hunk 1 above)
Comment (domagojlatecki) [reply src/app/admin/admin-campaign-new/admin-campaign-new.component.ts:285]: True, I will fix that 😄

**Review by mightymatth — CHANGES_REQUESTED** [review]

File: src/app/admin/admin-ft-asset-campaign-view/admin-ft-asset-campaign-view.component.html [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+      </ng-container>
+    </ng-container>
+
+    <ng-container *ngIf="campaignRes.error || assetRes.error">
+      <div class="w-full text-center mt-8">Cannot fetch campaign.</div>
+      <div hidden>Campaign errors: {{ campaignRes.error | json }}</div>
+      <div hidden>Asset errors: {{ assetRes.error | json }}</div>
+    </ng-container>
+
+    <ng-container *ngIf="campaignRes.loading && assetRes.loading">
```
[the remark this answers is not in the document]
Comment (domagojlatecki) [reply src/app/admin/admin-ft-asset-campaign-view/admin-ft-asset-campaign-view.component.html]: Using `campaignRes.loading || assetRes.loading` can cause both text error and loading to be visible at the same time. I wanted this behavior:
- if both asset and campaign are successfully loaded, show the screen
- if either asset or campaign got an error, show error (even if the other is still loading or successfully loaded)
- if something is still loading, then show loading
What I should have done here is this: `(campaignRes.loading || assetRes.loading) && !campaignRes.error && !assetRes.error` - that would probably do what I want, if there is at least one error show error screen even if something is loading. And if something is successful and the other thing is still loading, then show loading.

**Pull request merged**